07-routing-gossip.md

```diff
@@ -1119,7 +1120,59 @@ A->D's `update_add_htlc` message would be:
 And D->C's `update_add_htlc` would again be the same as B->C's direct payment
 above.
 
-## References
+# Onion Messages
+
+Onion messages allow peers to use existing connections to query for
+invoices (see [BOLT 12](12-offer-encoding.md)).  Like gossip messages,
+they are not associated with a particular local channel.  Like HTLCs,
+they use [BOLT 4](04-onion-routing.md#onion-messages) protocol for
+end-to-end encryption.
+
+Onion messages are unreliable: in particular, they are designed to
+be cheap to process and require no storage to forward.  As a result,
+there is no error returned from intermediary nodes.
+
+To enable messaging via blinded paths, there is an optional `blinding`
+parameter which allows decryption of the `enctlv` field inside the
+`onionmsg`'s `onionmsg_payload`.
+
+## The `onion_message` Message
+
+1. type: 387 (`onion_message`) (`option_onion_messages`)
+2. data:
+    * [`point`:`blinding`]
+    * [`u16`:`len`]
+    * [`len*byte`:`onionmsg`]
+
+## Requirements
+
+The writer:
+- MUST populate the per-hop payloads as described in [BOLT 4](04-onion-routing.md#onion-messages).
+- SHOULD retry via a different route if it expects a response and
+  doesn't receive one after a reasonable period.
+- SHOULD set `len` to 1366 or 32834.
```

Comment:
In BOLT 4, `hop_payloads` is said to be of length 1300. Shouldn't it be `len - 66` for onion messages?

Reply:
I believe the length here *includes* the MAC, version byte, and ephemeral public key, so its 1300 + 66 in the constant overhead. Again, all of this would be 10x more clear if we just specified the full messages instead of trying to make references.

Reply:
I understand that there is a constant overhead of 66 bytes, what I'm saying is that in BOLT 4, it says that `hop_payloads` is **always** of length 1300 but I don't think it should be true in the case of onion messages. If the `len` here is equal to 32834, then the `hop_payloads` should be of length 32768, not 1300. That should be mentioned in BOLT 4 (I wanted to put my comment there but Github wouldn't let me comment on a line that was not modified).

Reply:
Ah, then I think we're agreeing - we should just define the messages instead of making a reference to another (already defined) message and trying to indicate the changes to the other message.

Reply:
The `onionmsg` contains 66 bytes of version, public key and hmac. So if `len` is 32834, then the `hop_payloads` would be of length 32768. I guess that's where it comes from but as I see it it's just a big arbitrary number.

Reply:
Of course, I'm dumb, I missed the 66 bytes.
As long as it's a `SHOULD`, not a `MUST`, I just wanted to understand the rationale, so it explains it thanks :+1: